File: ColorSpaces/interactiveColorDetect.cpp
#include "color_spaces.hpp"

#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

#include <algorithm>
#include <filesystem>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>

#ifndef COLOR_SPACES_PROJECT_DIR
#error "COLOR_SPACES_PROJECT_DIR must be defined by CMake"
#endif

namespace {

constexpr int kPanelWidth = 420;
const std::string kWindowName = "Move the mouse over the image; press Esc to exit";

int parseInteger(const std::string& text) {
    std::size_t consumed = 0;
    const int value = std::stoi(text, &consumed);
    if (consumed != text.size()) {
        throw std::invalid_argument("invalid integer: " + text);
    }
    return value;
}

struct Options {
    std::filesystem::path input =
        std::filesystem::path(COLOR_SPACES_PROJECT_DIR) / "images" / "rub00.jpg";
    std::optional<std::filesystem::path> output;
    std::optional<int> x;
    std::optional<int> y;
    bool noDisplay = false;
    bool validate = false;
    bool help = false;
};

Options parseArguments(const int argc, char** argv) {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string argument(argv[index]);
        if (argument == "--input") {
            if (++index >= argc) {
                throw std::invalid_argument("--input requires a path");
            }
            options.input = std::filesystem::absolute(argv[index]);
        } else if (argument == "--output") {
            if (++index >= argc) {
                throw std::invalid_argument("--output requires a path");
            }
            options.output = std::filesystem::absolute(argv[index]);
        } else if (argument == "--x") {
            if (++index >= argc) {
                throw std::invalid_argument("--x requires an integer");
            }
            options.x = parseInteger(argv[index]);
        } else if (argument == "--y") {
            if (++index >= argc) {
                throw std::invalid_argument("--y requires an integer");
            }
            options.y = parseInteger(argv[index]);
        } else if (argument == "--no-display") {
            options.noDisplay = true;
        } else if (argument == "--validate") {
            options.validate = true;
        } else if (argument == "--help" || argument == "-h") {
            options.help = true;
        } else {
            throw std::invalid_argument("unknown argument: " + argument);
        }
    }
    return options;
}

cv::Mat renderPixelPanel(const cv::Mat& image, const int x, const int y) {
    if (x < 0 || x >= image.cols || y < 0 || y >= image.rows) {
        throw std::invalid_argument(
            "pixel coordinate is outside image bounds " +
            std::to_string(image.cols) + "x" + std::to_string(image.rows));
    }

    const color_spaces::PixelValues values =
        color_spaces::convertPixel(image.at<cv::Vec3b>(y, x));
    cv::Mat panel = cv::Mat::zeros(image.rows, kPanelWidth, CV_8UC3);
    const std::array<std::string, 5> lines = {
        "Pixel (x=" + std::to_string(x) + ", y=" + std::to_string(y) + ")",
        "BGR   " + color_spaces::vectorText(values.bgr),
        "HSV   " + color_spaces::vectorText(values.hsv),
        "YCrCb " + color_spaces::vectorText(values.yCrCb),
        "Lab   " + color_spaces::vectorText(values.lab),
    };

    for (std::size_t index = 0; index < lines.size(); ++index) {
        cv::putText(
            panel,
            lines[index],
            cv::Point(20, 45 + static_cast<int>(index) * 55),
            cv::FONT_HERSHEY_SIMPLEX,
            0.75,
            cv::Scalar(255, 255, 255),
            2,
            cv::LINE_AA);
    }

    cv::Mat marked = image.clone();
    cv::drawMarker(
        marked,
        cv::Point(x, y),
        cv::Scalar(0, 255, 255),
        cv::MARKER_CROSS,
        18,
        2);
    cv::Mat combined;
    cv::hconcat(marked, panel, combined);
    return combined;
}

void validateKnownPixel() {
    const color_spaces::PixelValues values =
        color_spaces::convertPixel(cv::Vec3b(40, 158, 16));
    if (values.bgr != cv::Vec3b(40, 158, 16) ||
        values.hsv != cv::Vec3b(65, 229, 158) ||
        values.yCrCb != cv::Vec3b(102, 67, 93) ||
        values.lab != cv::Vec3b(145, 71, 177)) {
        throw std::runtime_error("known pixel conversion regression");
    }
}

void runValidation(const Options& options) {
    validateKnownPixel();
    const cv::Mat image = color_spaces::readBgr(options.input);
    const cv::Mat rendered =
        renderPixelPanel(image, image.cols / 2, image.rows / 2);
    if (rendered.rows != image.rows || rendered.cols != image.cols + kPanelWidth) {
        throw std::runtime_error("rendered panel dimensions are incorrect");
    }
    if (options.output.has_value()) {
        color_spaces::writeImage(*options.output, rendered);
    }
    std::cout << "VALIDATION PASSED: image=" << image.cols << 'x' << image.rows
              << ", panel=" << rendered.cols << 'x' << rendered.rows << '\n';
}

struct MouseState {
    const cv::Mat* image = nullptr;
    cv::Mat rendered;
};

void onMouse(const int event, const int x, const int y, int, void* userData) {
    if (event != cv::EVENT_MOUSEMOVE || userData == nullptr) {
        return;
    }
    auto* state = static_cast<MouseState*>(userData);
    if (state->image != nullptr &&
        x >= 0 && x < state->image->cols &&
        y >= 0 && y < state->image->rows) {
        state->rendered = renderPixelPanel(*state->image, x, y);
        cv::imshow(kWindowName, state->rendered);
    }
}

void printUsage(const char* program) {
    std::cout
        << "Usage: " << program
        << " [--input IMAGE] [--x X --y Y] [--output IMAGE]"
        << " [--no-display] [--validate]\n";
}

}  // namespace

int main(const int argc, char** argv) {
    try {
        const Options options = parseArguments(argc, argv);
        if (options.help) {
            printUsage(argv[0]);
            return 0;
        }
        if (options.validate) {
            runValidation(options);
            return 0;
        }

        const cv::Mat image = color_spaces::readBgr(options.input);
        const int x = options.x.value_or(image.cols / 2);
        const int y = options.y.value_or(image.rows / 2);
        cv::Mat rendered = renderPixelPanel(image, x, y);
        const color_spaces::PixelValues values =
            color_spaces::convertPixel(image.at<cv::Vec3b>(y, x));

        std::cout << "Input: " << options.input << '\n';
        std::cout << "Pixel: x=" << x << ", y=" << y << '\n';
        std::cout << "BGR: " << color_spaces::vectorText(values.bgr) << '\n';
        std::cout << "HSV: " << color_spaces::vectorText(values.hsv) << '\n';
        std::cout << "YCrCb: " << color_spaces::vectorText(values.yCrCb) << '\n';
        std::cout << "Lab: " << color_spaces::vectorText(values.lab) << '\n';

        if (options.output.has_value()) {
            color_spaces::writeImage(*options.output, rendered);
            std::cout << "Wrote: " << *options.output << '\n';
        }
        if (options.noDisplay) {
            return 0;
        }

        MouseState state{&image, rendered};
        cv::namedWindow(kWindowName, cv::WINDOW_AUTOSIZE);
        cv::setMouseCallback(kWindowName, onMouse, &state);
        cv::imshow(kWindowName, state.rendered);
        while ((cv::waitKey(20) & 0xFF) != 27) {
        }
        cv::destroyAllWindows();
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        printUsage(argv[0]);
        return 2;
    }
}

File: ColorSpaces/interactiveColorDetect.py
#!/usr/bin/env python3
"""Inspect a BGR pixel in HSV, YCrCb, and Lab encodings."""

from __future__ import annotations

import argparse
from collections.abc import Sequence
from pathlib import Path

import cv2
import numpy as np

from color_spaces import (
    convert_pixel,
    read_bgr,
    resolve_input_path,
    run_core_validation,
    write_image,
)

WINDOW_NAME = "Move the mouse over the image; press Esc to exit"


def render_pixel_panel(image: np.ndarray, x: int, y: int) -> np.ndarray:
    """Render the image beside a text panel describing one valid pixel."""

    height, width = image.shape[:2]
    if not (0 <= x < width and 0 <= y < height):
        raise ValueError(f"pixel ({x}, {y}) is outside image bounds {width}x{height}")

    values = convert_pixel(image[y, x])
    panel = np.zeros((height, 420, 3), dtype=np.uint8)
    lines = [
        f"Pixel (x={x}, y={y})",
        f"BGR   {values['BGR']}",
        f"HSV   {values['HSV']}",
        f"YCrCb {values['YCrCb']}",
        f"Lab   {values['Lab']}",
    ]
    for index, text in enumerate(lines):
        cv2.putText(
            panel,
            text,
            (20, 45 + index * 55),
            cv2.FONT_HERSHEY_SIMPLEX,
            0.75,
            (255, 255, 255),
            2,
            cv2.LINE_AA,
        )

    marked = image.copy()
    cv2.drawMarker(marked, (x, y), (0, 255, 255), cv2.MARKER_CROSS, 18, 2)
    return np.hstack((marked, panel))


def run_validation(output: str | Path | None = None) -> dict[str, int]:
    """Run deterministic dataset and rendering checks."""

    metrics = run_core_validation()
    image = read_bgr(resolve_input_path(None))
    x_coordinate = image.shape[1] // 2
    y_coordinate = image.shape[0] // 2
    rendered = render_pixel_panel(image, x_coordinate, y_coordinate)
    if rendered.shape != (image.shape[0], image.shape[1] + 420, 3):
        raise AssertionError(f"unexpected rendered shape: {rendered.shape}")
    if output is not None:
        write_image(output, rendered)

    print(
        "VALIDATION PASSED: "
        f"{metrics['cube_images']} cube images, "
        f"{metrics['piece_images']} piece images, "
        f"panel={rendered.shape[1]}x{rendered.shape[0]}"
    )
    return {
        **metrics,
        "output_width": rendered.shape[1],
        "output_height": rendered.shape[0],
    }


def _build_parser() -> argparse.ArgumentParser:
    """Create the command-line parser."""

    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", type=Path, help="input image; defaults to images/rub00.jpg")
    parser.add_argument("--x", type=int, help="sample x coordinate; defaults to image center")
    parser.add_argument("--y", type=int, help="sample y coordinate; defaults to image center")
    parser.add_argument("--output", type=Path, help="optional annotated output image")
    parser.add_argument("--no-display", action="store_true", help="do not open a GUI window")
    parser.add_argument("--validate", action="store_true", help="run deterministic checks")
    return parser


def main(argv: Sequence[str] | None = None) -> int:
    """Run the CLI or interactive pixel inspector."""

    arguments = _build_parser().parse_args(argv)
    if arguments.validate:
        run_validation(arguments.output)
        return 0

    image_path = resolve_input_path(arguments.input)
    image = read_bgr(image_path)
    x_coordinate = arguments.x if arguments.x is not None else image.shape[1] // 2
    y_coordinate = arguments.y if arguments.y is not None else image.shape[0] // 2
    rendered = render_pixel_panel(image, x_coordinate, y_coordinate)

    values = convert_pixel(image[y_coordinate, x_coordinate])
    print(f"Input: {image_path}")
    print(f"Pixel: x={x_coordinate}, y={y_coordinate}")
    for name, value in values.items():
        print(f"{name}: {value}")

    if arguments.output is not None:
        output_path = write_image(arguments.output, rendered)
        print(f"Wrote: {output_path}")

    if arguments.no_display:
        return 0

    state = {"image": image, "rendered": rendered}

    def on_mouse(event: int, x_value: int, y_value: int, _flags: int, _data: object) -> None:
        if event != cv2.EVENT_MOUSEMOVE:
            return
        if 0 <= x_value < image.shape[1] and 0 <= y_value < image.shape[0]:
            state["rendered"] = render_pixel_panel(image, x_value, y_value)
            cv2.imshow(WINDOW_NAME, state["rendered"])

    cv2.namedWindow(WINDOW_NAME, cv2.WINDOW_AUTOSIZE)
    cv2.setMouseCallback(WINDOW_NAME, on_mouse)
    cv2.imshow(WINDOW_NAME, state["rendered"])
    while cv2.waitKey(20) & 0xFF != 27:
        pass
    cv2.destroyAllWindows()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
